Pr for migration endpoints

benpate opened the pull request:
This addresses #42 by moving new collections into the `migration` property in the Actor's profile, along with some explanatory text about how these collections behave.

I added one note to the "Open Issues" section: I believe the issue of negotiating extended data formats is addressed by the ability to define custom collections within the `migration` property. I've left this lightly defined right now, awaiting your feedback. Ideally, we should define a strict process that uses `@context` property to declare these app-specific collections. If you'd like, just say "go" and I'll take a shot at this, too.

As always, please let me know if there's anything I should change and I'll happily make edits.

benpate committed:
Represent "content" and "migration" within a separate property

Also adding some explanatory text to the "Migration Outbox" and "Content Collection" sections.  I think this could be better, but it's a good start.

benpate committed:
"Tombstone" is an object not an activity

minor change to correctly represent "Tombstone" as an object, not as an activity.

TallTed commented on lola.html:
```diff
@@ -635,29 +639,45 @@ <h3>Authorization</h3>
       <h3>Migration Outbox</h3>
+	  <p>
+		The migration outbox is intended as a simple way for applications to enable LOLA account migration.  Servers implementing LOLA MUST include this collection in actor's profiles, as it will provide maximum compatibility with other ActivityPub software.
+		This collection SHOULD contain a filtered view of all the actor's activities that the source server deems eligible for migration.  The source server MAY filter out activities that are not relevant to the destination server, or that the user has chosen to exclude during the OAuth authorization step.
```
Remember that HTML source formatting doesn't get reflected in the HTML rendering. If it's worth breaking paragraphs in the source, it's probably worth breaking them for the rendering.
```suggestion
		The migration outbox is intended as a simple way for applications to enable LOLA account migration.  Servers implementing LOLA MUST include this collection in actor profiles, as it will provide maximum compatibility with other ActivityPub software.
	  </p>
      <p>
        This collection SHOULD contain a filtered view of all the actor's activities that the source server deems eligible for migration.  The source server MAY filter out activities that are not relevant to the destination server, or that the user has chosen to exclude during the OAuth authorization step.
```
(It likely makes sense to keep all lines of the HTML source ~80 chars or less, but that's a separate consideration.)

benpate replied:
Sure, this makes sense.  I usually just view the document with "word wrap", but I can follow whatever standard is in place here.

Speaking of HTML code formatting, there's a lot of small inconsistencies throughout the document.  Further down the road, I was planning to ask if some sort of automated code formatter might be useful. Since we're here, I'm mentioning it now - would that be useful to this project?

TallTed commented on lola.html:
```diff
@@ -635,29 +639,45 @@ <h3>Authorization</h3>
       <p>
-        The destination server MUST use the migration outbox in lieu of the regular outbox if the migration outbox URL 
-        is different.  This allows the source server some flexibility that may be important, in how it fetches and 
+        When migrating user's data, the destination server MUST use the migration outbox in lieu of the regular outbox if the migration outbox URL is different.  This allows the source server some flexibility that may be important, in how it fetches and 
```
```suggestion
        When migrating user data, the destination server MUST use the migration outbox in lieu of the regular outbox if the migration outbox URL is different.  This allows the source server some flexibility that may be important, in how it fetches and 
```

TallTed commented on lola.html:
```diff
@@ -635,29 +639,45 @@ <h3>Authorization</h3>
         </li>
         <li>
-          Media objects that are hosted by the source server - if it is a resource that would go away if the source 
+          Media objects (such as images, videos, or audio files) that are hosted by the source server - if it is a resource that would go away if the source 
```
```suggestion
          Media objects (such as image, video, or audio files) that are hosted by the source server — if it is a resource that would go away if the source 
```

TallTed commented on lola.html:
```diff
@@ -693,7 +713,7 @@ <h3>Content Collection</h3>
 
       <p>
-        Open Issues:  Is there a way to negotiate what extended features are recognized and whether the source should downgrade structured info?  Is there already a way to request posts in one format or another? 
+        Open Issues:  Is there a way to negotiate what extended features are recognized and whether the source should downgrade structured info?  Is there already a way to request posts in one format or another? -- Perhaps this is addressed by the ability to defined additional collections in the "migration" property.
```
```suggestion
        Open Issues:  Is there a way to negotiate the extended features that are recognized and/or whether the source should downgrade structured info?  Is there already a way to request posts in one format or another? — Perhaps this is addressed by the ability to define additional collections in the "migration" property?
```

TallTed commented on lola.html:
```diff
@@ -761,7 +781,7 @@ <h4>Like, Follow and Block Activities</h4>
       <p>
-        Add, Update, Undo, Remove, Tombstone and Delete activities should not be copied.  Rather than propagate these, the 
+        Add, Update, Undo, Remove and Delete activities should not be copied.  Rather than propagate these, the 
         results of changes SHOULD be provided by the source server.  
```
```suggestion
        Add, Update, Undo, Remove, and Delete activities should not be copied.  Rather than propagate these, the
        source server SHOULD provide the <em>results</em> of changes.
```

TallTed commented on lola.html:
```diff
@@ -773,7 +793,7 @@ <h4>Change Type Activities</h4>
       <p>
-        Another example, if a server posts a Note then replaces that with a Tombstone activity, the Tombstone activity 
+        Another example, if a server posts a Note then replaces that with a Tombstone object, the Tombstone object 
         SHOULD be omitted by the source and SHOULD NOT be copied by the destination.  
```
```suggestion
        Another example: If a server posts a Note and then replaces that with a Tombstone object, the Tombstone object
        SHOULD be omitted by the source and SHOULD NOT be copied by the destination.
```

TallTed reviewed (changes requested):
Editorial, grammar and the like.

benpate committed:
Format HTML to 80 characters wide

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

benpate committed:
Format HTML to 80 characters wide

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

benpate committed:
Update lola.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

benpate committed:
Update lola.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

benpate committed:
Editing for clarity

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

benpate committed:
Grammatical fixes

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

lisad commented on lola.html:
```diff
@@ -507,17 +507,21 @@ <h3>Feature Discovery</h3>
     "liked": "https://oakfrost.example.com/brock/liked",
     "blocked": "https://oakfrost.example.com/brock/blocked",
-    "migration": "https://oakfrost.example.com/brock/outbox"
+    "migration": {
```
Since "following" and "blocked" are important for migration, should these URLs also be copied (or moved) into the "migration" collection?  From what I've seen, servers aren't already putting those collections on the Actor object - the Actor object is normally accessed by other servers in the fediverse, and the way the user adds following and blocked accounts is for example using Mastodon's client API instead of ActivityPub.  But I made sure these are required for LOLA.

benpate replied:
Different servers do different things, so it would have to depend on which specific app we're talking about :(

You could make a case for duplicating any available collection into the actor's "migration" property, which would signal affirmatively that "this collection is safe to migrate".  If the server intends to just use the regular collection, then it would be really easy to just duplicate the original ID under "migration" too.  Plus, this would give target servers a single place to look for all of their data.

I'm super new here, so I don't want to rock the boat without checking in before I make any big changes. But if this looks good to you, I can certainly rewrite this PR so that "following" and "blocked" are also supposed to be sourced from the actor's "migration" property.

lisad replied:
I can see this one equally either way. Obviously the block/follow collections may be used to populate a new account if the new account server can handle the functionality and the user agrees to do so, whether or not they're in the migration property.  Just leaving them where they are for regular use as well as migration use is less fuss; adding them to the migration property is more clear.  So I don't care which way we go on this one and if you have a stronger opinion we can go with that.

When modifying protocol docs I just want to group changes together and try to make them complete - better to have our small set of LOLA implementors so far make 1 change to add "migration" property than have to modify it again later too.

benpate replied:
(sorry it took me a bit to respond here)

Yes. Fewer changes to the design doc are better. And there's no rush to merge my changes until everyone is happy with them.

If `Following` and `Blocked` are available for migration, then I think I'll recommend putting them in this property as well.  Like you said, servers are already inconsistent about what they display in Actor's profiles, so I think it would be good to make the `migration` section as clear and well-defined as possible.

And if a server has special requirements, they can always define a custom schema and put that in the `migration` property as well.

lisad replied:
Yes I agree with the argument that copying collections that the source things are important for migrations into the migrations property is useful. It's more clear than just having them in one or the other place.  Why don' t you merge with or without that change, and we can copy block/follow collection URLs in there too.

benpate replied:
Awesome.  Will do.

benpate replied:
@lisad I've just (finally) posted this change.  Thanks for being patient with me :)

benpate committed:
Update document to include "following" and "blocked" properties

I've made the changes discussed in this PR, including the "blocked" and "following" collections in the "migration" property, along with some (hopefully better) explanations of what these are doing.

Please let me know anything I should change/add/delete from this.
